torch/autograd/profiler.py

```diff
@@ -458,15 +459,19 @@ def _call_end_callbacks_on_future(self, fut: Future[Any]) -> Future[Any]:
 
         """
         # Throw if we have already attached a callback onto the future.
-        if not self.run_callbacks_on_exit:
+        if len(self.rec) == 0:
             raise RuntimeError("_call_end_callbacks_on_future can only be called once.")
 
         # We are scheduling to run this RecordFunction's end callbacks when the
-        # passed in future completes, so don't run end callbacks on exit.
-        self.run_callbacks_on_exit = False
-        profiled_future = torch.ops.profiler._call_end_callbacks_on_jit_fut(self.handle, fut)
+        # passed in future completes, so don't run end callbacks on exit. The
+        # future will capture the RecordFunction object and extend its lifetime.
+        profiled_future = torch.ops.profiler._call_end_callbacks_on_jit_fut(self.rec.pop(), fut)
         return profiled_future
 
+    @property
+    def handle(self):
```

Comment:
Do we need to add error checking to ensure this list isn't empty?

Reply:
The `RecordFunction` object originally hold by this context manager is called `handle`. This property exists is purely due to existing tests use this internal object for testing.

Anyway, if we add error checking, do we simply capture the `IndexError` and reraise it with clearer `RuntimeError`?

Reply:
I added an error checking with comprehensive `RuntimeError` msg.